infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/provision/AsyncStoragePodWatcher.java

```diff
@@ -0,0 +1,135 @@
+import org.eclipse.che.commons.schedule.ScheduleDelay;
+import org.eclipse.che.workspace.infrastructure.kubernetes.KubernetesClientFactory;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+@Singleton
+public class AsyncStoragePodWatcher {
+
+  private static final Logger LOG = LoggerFactory.getLogger(AsyncStoragePodWatcher.class);
+
+  private final KubernetesClientFactory kubernetesClientFactory;
+  private final UserManager userManager;
+  private final PreferenceManager preferenceManager;
+  private final WorkspaceRuntimes runtimes;
+  private final long shutdownTimeoutSec;
+  private final Clock clock = Clock.systemDefaultZone();
+
+  @Inject
+  public AsyncStoragePodWatcher(
+      KubernetesClientFactory kubernetesClientFactory,
+      UserManager userManager,
+      PreferenceManager preferenceManager,
+      WorkspaceRuntimes runtimes,
+      @Named("che.infra.kubernetes.async.storage.shutdown_timeout_min") long shutdownTimeoutMin) {
+    this.kubernetesClientFactory = kubernetesClientFactory;
+    this.userManager = userManager;
+    this.preferenceManager = preferenceManager;
+    this.runtimes = runtimes;
+    this.shutdownTimeoutSec = shutdownTimeoutMin * 60;
+  }
+
+  @ScheduleDelay(
+      unit = TimeUnit.MINUTES,
+      initialDelay = 1,
+      delayParameterName = "che.infra.kubernetes.async.storage.shutdown_check_period_min")
+  public void check() {
+    for (User user :
+        iterateLazily((maxItems, skipCount) -> userManager.getAll(maxItems, skipCount))) {
+      try {
+        String owner = user.getId();
+        Map<String, String> preferences = preferenceManager.find(owner);
+        String lastTimeAccess = preferences.get(LAST_ACTIVITY_TIME);
+        String namespace = preferences.get(LAST_ACTIVE_INFRASTRUCTURE_NAMESPACE);
+        if (isNullOrEmpty(namespace)
+            || isNullOrEmpty(lastTimeAccess)
+            || isAnyRuntimeInProgress(owner)) {
+          continue;
+        }
+        long lastTimeAccessSec = parseLong(lastTimeAccess);
+        long epochSec = clock.instant().getEpochSecond();
+        if (epochSec - lastTimeAccessSec >= shutdownTimeoutSec) {
+          PodResource<Pod, DoneablePod> podDoneablePodPodResource =
+              kubernetesClientFactory
+                  .create()
+                  .pods()
+                  .inNamespace(namespace)
+                  .withName(ASYNC_STORAGE);
+          if (podDoneablePodPodResource.get() != null) {
+            podDoneablePodPodResource.delete();
+          }
+        }
+      } catch (InfrastructureException | ServerException e) {
+        LOG.error(e.getMessage(), e);
+      }
+    }
+  }
+
+  /**
+   * Going to check is currently owner has workspaces in progress: it's status is {@link
+   * org.eclipse.che.api.core.model.workspace.WorkspaceStatus#STARTING} or {@link
+   * org.eclipse.che.api.core.model.workspace.WorkspaceStatus#STOPPING})
+   *
+   * @param owner the user id to check
+   */
```

Comment:
maybe it make sense to have such a method in WorkspaceRuntimes.
Need to here other opinions. @sleshchenko @sparkoo @amisevsk @metlos

Reply:
small +1 from me to move it in WorkspaceRuntimes.